### PR TITLE
s390x: use runtime detection for vectorization support

### PR DESCRIPTION
**Pull request by AlekseiNikiforovIBM**
s390x: use runtime detection for vectorization support

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: aten/src/ATen/native/DispatchStub.cpp [hunk 1]
```diff
@@ -10,8 +10,39 @@
...
+#include <regex>
+#include <string>
+#endif
+
 namespace at::native {
 
+#ifdef HAVE_ZVECTOR_CPU_DEFINITION
+static bool proc_cpuinfo_field_contains(const std::string &field_name, const std::string &value)
+{
+  std::ifstream cpuinfofile("/proc/cpuinfo");
```
Comment (jansel) [inline aten/src/ATen/native/DispatchStub.cpp:24]: How portable is this?  What about Windows?

**Review by malfet — APPROVED** [review]
This is probably OK, but it feels like a more straightforward path would be to enable this detection in https://github.com/pytorch/cpuinfo and query from PyTorch

File: aten/src/ATen/native/DispatchStub.cpp [hunk 2]
```diff
@@ -60,10 +91,16 @@ static CPUCapability compute_cpu_capability() {
 #endif
   }
 #endif
+
+#ifdef HAVE_ZVECTOR_CPU_DEFINITION
+  // vxe is needed for fp32 vector instructions
+  if (proc_cpuinfo_field_contains("features", "vxe")) {
```
[the remark this answers is not in the document]
Comment (AlekseiNikiforovIBM) [reply aten/src/ATen/native/DispatchStub.cpp:97]: This function is called once and result is stored here:

https://github.com/pytorch/pytorch/pull/123936/files/ff4e6ea089212849eadf599da4ef96a86adc83fd#diff-54c373491da67eb31c3777457d7b043a49dd3966412edfd928ffd2013e4d6a54R109-R112

File: aten/src/ATen/native/DispatchStub.cpp [hunk 3]
```diff
@@ -10,8 +10,39 @@
...
+#endif
+
 namespace at::native {
 
+#ifdef HAVE_ZVECTOR_CPU_DEFINITION
+static bool proc_cpuinfo_field_contains(const std::string &field_name, const std::string &value)
+{
+  std::ifstream cpuinfofile("/proc/cpuinfo");
+  std::string line;
+  std::regex feature_regex(std::string("^") + field_name + std::string("[[:space:]]*:[[:space:]]*(.*)$"));
```
[the remark this answers is not in the document]
Comment (Andreas-Krebbel) [reply aten/src/ATen/native/DispatchStub.cpp]: The preferred way to check for these kind of features is the getauxval glibc function. You can use it to query the hwcaps flags passed from the kernel to user space as part of the auxiliary vector.

**Review by jansel — APPROVED** [review]

File: aten/src/ATen/native/DispatchStub.cpp (hunk 1 above)
Comment (xuhancn) [reply aten/src/ATen/native/DispatchStub.cpp]: Windows seems not have any version for IBM CPU.

**Review by xuhancn — APPROVED** [review]

**Pull request closed without merging**